[Belt] Add List.sort

zploskey opened the pull request:
Implements Belt.List.sort as specified in the issue.
Adds a couple of simple tests and an example.

Fixes #2586.

zploskey commented:
Apparently [Array.prototype.sort()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort) sorts everything like a string, and is what `Js.Array.sortInPlace()` is bound to. Although I could easily fix the test, I don't think that function is what we want to use here. It's unintuitive for a sort function give a result like `[1; 3; 3; 37; 9]`. Maybe better to use `sortInPlaceWith(function a b -> a - b)`.

chenglou commented:
Oh no, definitely don't use Js.Array.sortInPlace. That behavior is famously confusing. Maybe `SortArray.stableSortInPlaceBy`? cc @bobzhang

zploskey commented:
I went ahead and adjusted it to use `SortArray.stableSortInPlaceBy`. I figure whatever solution we settle on will use a compare function.

chenglou commented:
The asymmetry between SortArray and List.sort is a little weird. I'm wondering how we should arrange these

zploskey commented:
What was the reasoning behind putting Sort functions on arrays in their own separate module?

bobzhang commented on jscomp/others/belt_List.ml:
```diff
@@ -692,6 +692,13 @@ let setAssocU xs x k eq =
 let setAssoc xs x k eq = setAssocU xs x k (fun [@bs] a b -> eq a b)
 
+let sort xs cmp =
+  let arr = ref (toArray xs) in
```
why a ref needed here

bobzhang replied:
sortU/sort

zploskey replied:
Ok, I'll add a sortU version. I think I was confused about whether I could use let and it still be mutated by the in-place sort. It works fine without the ref.

bobzhang commented on jscomp/others/belt_List.ml:
```diff
@@ -692,6 +692,13 @@ let setAssocU xs x k eq =
+let sort xs cmp =
+  let arr = ref (toArray xs) in
+  begin
+    Belt_SortArray.stableSortInPlaceBy !arr cmp;
```
A.stableSortInPlaceBy -- it is already aliased in the beginning

zploskey replied:
Belt_Array is aliased as A in this module. It doesn't have the sort functions from Belt_SortArray. Should Belt_SortArray be opened in Belt_Array?

bobzhang replied:
This is fine then, sorry my mistake

zploskey commented:
I addressed the concerns and rebased on master so that all the tests run. I'm not sure why we're seeing a build failure on Travis. It builds fine for me locally. Somehow it can't find `Belt_SortArray.cmj`, but it just built it a few lines before. :man_shrugging:

zploskey commented:
That seems to have worked. Let me know if there are any other concerns.

bobzhang commented on jscomp/others/belt_List.ml:
```diff
@@ -692,6 +692,16 @@ let setAssocU xs x k eq =
+  let arr = toArray xs in
+  Belt_SortArray.stableSortInPlaceByU arr cmp;
+  fromArray arr
+
```
In general, in belt, we should define `sortU` first and wrap `sort` like this

```ocaml
let sort xs cmp = sortU xs (fun[@bs] x y -> cmp x y)
```
The thing is in your `sort`, you are calling `stableSortInPlaceBy` instead of `stableSortInPlaceByU` which is slower

zploskey replied:
Fixed. Thanks!
